### DISCUSSION TITLE
Dependant providers and using WidgetRef from outside widget tree

### DISCUSSION THREAD

**Discussion by emirhalici** [rrousselGit/riverpod#2364] [Q&A]
Is it okay to use WidgetRef to access other providers? A really common scenario I find myself in is this:

```
// Imagine this method is inside Provider Y
void doSomething() {
  // Some async computation
  double result = 1.000;

  if (result > 1) {
    // Provider X should do something
  } else {
    // Other logic
  }
}
```

This behavior can be done by passing in a WidgetRef from UI so that we now have access to the provider. How does WidgetRef lifecycles work? Do they live in the same lifespan of the Widget they get passed from? Could this approach break when UI element gets demounted before WidgetRef gets used?

You may think to combine states to solve this problem but another problem may arise there: What exactly happens when recomputation is triggered for a provider WHILE it is doing some async operation? Say provider was retrieving something from an api or waiting for a service to do it. In between this process, parent dependencies changed. Dirty provider completed execution after but it is not valid anymore.

**Comment by rrousselGit** [comment]
Using WidgetRef is equivalent to putting your business logic in the UI. You should consider refactoring it to use Ref. 
So no this isn't recommended :)

**Reply by emirhalici** [reply]
Then my question would be about provider creation callback. 

When listened values change over time, does entire provider get replaced in favor of the new provider instance? 
For example: When a stateless widget's parameters change entire thing is replaced with a new one. When a stateful widget's parameters change it calls didUpdateWidget and doesn't delete the old instance. 
Does Riverpod providers have such mechanisms?
